libraries/architecture/src/test/kotlin/io/element/android/libraries/architecture/AsyncKtTest.kt

```diff
@@ -0,0 +1,113 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.libraries.architecture
+
+import androidx.compose.runtime.MutableState
+import com.google.common.truth.Truth.assertThat
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.test.runTest
+import org.junit.Test
+
+class AsyncKtTest {
+    @Test
+    fun `updates state when block returns success`() = runTest {
+        val state = TestableMutableState<Async<Int>>(Async.Uninitialized)
+
+        val result = runUpdatingState(state) {
+            delay(1)
+            Result.success(1)
+        }
+
+        assertThat(result.isSuccess).isTrue()
+        assertThat(result.getOrNull()).isEqualTo(1)
+
+        assertThat(state.pop()).isEqualTo(Async.Uninitialized)
+        assertThat(state.pop()).isEqualTo(Async.Loading(null))
+        assertThat(state.pop()).isEqualTo(Async.Success(1))
+        state.assertNoMoreValues()
+    }
+
+    @Test
+    fun `updates state when block returns failure`() = runTest {
+        val state = TestableMutableState<Async<Int>>(Async.Uninitialized)
+
+        val result = runUpdatingState(state) {
+            delay(1)
+            Result.failure(MyThrowable("hello"))
+        }
+
+        assertThat(result.isFailure).isTrue()
+        assertThat(result.exceptionOrNull()).isEqualTo(MyThrowable("hello"))
+
+        assertThat(state.pop()).isEqualTo(Async.Uninitialized)
+        assertThat(state.pop()).isEqualTo(Async.Loading(null))
+        assertThat(state.pop()).isEqualTo(Async.Failure<Int>(MyThrowable("hello")))
+        state.assertNoMoreValues()
+    }
+
+    @Test
+    fun `updates state when block returns failure transforming the error`() = runTest {
+        val state = TestableMutableState<Async<Int>>(Async.Uninitialized)
+
+        val result = runUpdatingState(state, { MyThrowable(it.message + " world") }) {
+            delay(1)
+            Result.failure(MyThrowable("hello"))
+        }
+
+        assertThat(result.isFailure).isTrue()
+        assertThat(result.exceptionOrNull()).isEqualTo(MyThrowable("hello world"))
+
+        assertThat(state.pop()).isEqualTo(Async.Uninitialized)
+        assertThat(state.pop()).isEqualTo(Async.Loading(null))
+        assertThat(state.pop()).isEqualTo(Async.Failure<Int>(MyThrowable("hello world")))
+        state.assertNoMoreValues()
+    }
+}
+
+/**
+ * A fake [MutableState] that allows to record all the states that were set.
+ */
+private class TestableMutableState<T>(
+    value: T
+) : MutableState<T> {
+
+    private val _deque = ArrayDeque<T>(listOf(value))
+
+    override var value: T
+        get() = _deque.last()
+        set(value) {
+            _deque.addLast(value)
+        }
+
+    /**
+     * Returns the states that were set in the order they were set.
+     */
+    fun pop(): T = _deque.removeFirst()
```

Comment:
Minor detail, but should this be `popFirst()`? I was a bit confused when I saw it worked beckwards to what I expected.

Reply:
I agree we want to add them as `last`, but if I call `pop()` on some collection, I expect it to return the last item added, not the first one, that's why I suggested using `popFirst()` instead, since I found it confusing: I had to actually check what this `pop` function was doing under the hood for the asserts to make sense.

Reply:
Oh, I now understand what you mean. Renamed to `popFirst()`.